devito/types/grid.py

```diff
@@ -505,7 +527,88 @@ def define(self, dimensions):
         return {d: ('middle', 1, 1) for d in dimensions}
 
 
-class SubDomainSet(SubDomain):
+class MultiSubDimension(SubDimension):
+
+    """
+    A special SubDimension for graceful lowering of MultiSubDomains.
+    """
+
+    def __init_finalize__(self, name, parent):
+        lst, rst = self._symbolic_thickness(name)
+        left = parent.symbolic_min + lst
+        right = parent.symbolic_max - rst
+
+        super().__init_finalize__(name, parent, left, right, ((lst, 0), (rst, 0)), False)
+
+
+class MultiSubDomain(SubDomain):
+
+    """
+    Base class for types representing a group of SubDomains.
+    """
+
+    @classmethod
+    def _bounds_glb_to_loc(cls, dec, m, M):
+        """
+        Translate a SubDomain global bounds, that is thicknesses, into local bounds.
+        """
+        bounds_m = np.zeros(m.shape, dtype=m.dtype)
+        bounds_M = np.zeros(m.shape, dtype=m.dtype)
+        for j in range(m.size):
+            lmin = dec.glb_min + m[j]
+            lmax = dec.glb_max - M[j]
+
+            # Check if the subdomain doesn't intersect with the decomposition
+            if lmin < dec.loc_abs_min and lmax < dec.loc_abs_min:
+                bounds_m[j] = dec.loc_abs_max
+                bounds_M[j] = dec.loc_abs_max
+                continue
+            if lmin > dec.loc_abs_max and lmax > dec.loc_abs_max:
+                bounds_m[j] = dec.loc_abs_max
+                bounds_M[j] = dec.loc_abs_max
+                continue
+
+            if lmin < dec.loc_abs_min:
+                bounds_m[j] = 0
+            elif lmin > dec.loc_abs_max:
+                bounds_m[j] = dec.loc_abs_max
+                bounds_M[j] = dec.loc_abs_max
+                continue
+            else:
+                bounds_m[j] = dec.index_glb_to_loc(m[j], LEFT)
+
+            if lmax < dec.loc_abs_min:
+                bounds_m[j] = dec.loc_abs_max
+                bounds_M[j] = dec.loc_abs_max
+                continue
+            elif lmax >= dec.loc_abs_max:
+                bounds_M[j] = 0
+            else:
+                bounds_M[j] = dec.index_glb_to_loc(M[j], RIGHT)
```

Comment:
I think this can simplify a bit 
```
           if lmin < dec.loc_abs_min:
                bounds_m[j] = 0
            elif lmin > dec.loc_abs_max:
                bounds_m[j] = dec.loc_abs_max
            else:
                bounds_m[j] = dec.index_glb_to_loc(m[j], LEFT)

            if lmax < dec.loc_abs_min:
                bounds_M[j] = dec.loc_abs_max
            elif lmax >= dec.loc_abs_max:
                bounds_M[j] = 0
            else:
                bounds_M[j] = dec.index_glb_to_loc(M[j], RIGHT)
```

and that should cover all the cases

Reply:
ah, this is just lifted from what is now a subclass -- this is code that @rhodrin wrote a time ago... did you see my ranting on slack about it? basically this would benefit from a rewrite regardless